### PR TITLE
[React] Use intersection types for key/ref; preserve instance type T on element types

### PR DESCRIPTION
**Pull request by vsiao**
Previously, you had to extend `React.Props<T>` (specifying instance type `T` explicitly) to allow callers to create elements with `key` and `ref` attributes. We now infer `T` automatically and utilize intersection types to accept `key` and `ref`:

``` ts
declare function createElement<P, T extends Component<P, {}>, C extends ComponentClass<P>>(
    type: C & (new() => T) & (new() => { props: P }),
    props?: P & { key?: Key; ref?: Ref<T> },
    ...children: ReactNode[]): ComponentElement<P, T>;
```

This PR introduces a new `React.ComponentElement<P, T>` type, which allows us to return the correct type `T` from `ReactDOM.render` (previously the best we could do was `Component<P, {}>`). I'm personally pretty excited about this.
## Breaking Changes
- `React.Props<T>` is now deprecated, which means that `key` and `ref` can't be accessed on `this.props` anymore. `children` is still available for the time being since we can't prevent it from being passed via `createElement` (h/t @adidahiya)
- `React.ReactHTMLElement` now takes a type parameter `<T extends HTMLElement>`
- `React.DOMElement` takes two type parameters `<P extends DOMAttributes, T extends Element>`
- same changes to `React.HTMLFactory` and `React.DOMFactory`
## New Interfaces

Attributes (analogous to `JSX.IntrinsicAttributes` and `JSX.IntrinsicClassAttributes<T>`):

``` ts
interface Attributes {
    key?: Key;
}
interface ClassAttributes<T> extends Attributes {
    ref?: Ref<T>;
}
```

Component Elements (aliased as `CElement`):

``` ts
interface ComponentElement<P, T extends Component<P, {}>> extends ReactElement<P> {
    type: ComponentClass<P>;
    ref?: Ref<T>;
}
```

Stateless Functional Component Elements (no `ref`):

``` ts
interface SFCElement<P> extends ReactElement<P> {
    type: StatelessComponent<P>;
}
```
## Other Changes
- added documentation to `README.md` on known problems and workarounds. I'd say `cloneElement` is the weakest link in these type definitions.
- made a number of `isXXX(...): boolean` API functions into custom typeguards
- added `nonce` and `reversed` HTML attributes to bring us up to date with React 0.14.4

cc @RyanCavanaugh @jbrantly @adidahiya

### CODE REVIEW

**Comment by stepancar** [comment]
:+1:

**Comment by stephenjelfs** [comment]
:+1: No problems with fixed-data-table changes.

**Comment by stkb** [comment]
:+1:

**Comment by pspeter3** [comment]
I'm really excited about this. This should make the type guarantees much better.

**Comment by isman-usoh** [comment]
:+1:

**Comment by vvakame** [comment]
:+1:

File: react/react.d.ts [hunk 1]
```diff
@@ -128,7 +163,13 @@ declare namespace __React {
         setState(state: S, callback?: () => any): void;
         forceUpdate(callBack?: () => any): void;
         render(): JSX.Element;
-        props: P;
+
+        // React.Props<T> is now deprecated, which means that the `children`
+        // property is not available on `P` by default, even though you can
+        // always pass children as variadic arguments to `createElement`.
+        // In the future, if we can define its call signature conditionallly
+        // on the existence of `children` in `P`, then we should remove this.
```
Comment (adidahiya) [inline react/react.d.ts]: :+1:

**Comment by progre** [comment]
@vsiao 
Nice change.
I try checkout this fix, and change my code. And then, I don't go anyhow well.
my code;
```
let component = ReactDOM.render(
    <MyComponent myProps={data}>,
    document.getElementById("target_id")) as MyComponent;
component.setState({ myState: data });
```

This fix changes component's type to uncertain. And It needs a cast.
I'm not too much detail to react. Can you tell me this in the correct one? and tell me if there is a more correct means?

**Comment by vsiao** [comment]
@progre unfortunately it seems like JSX users won't benefit from these changes directly: https://github.com/Microsoft/TypeScript/wiki/JSX#the-jsx-result-type

**Comment by progre** [comment]
I see.... thx @vsiao

**Comment by seansfkelley** [comment]
@vsiao is there a reason you changed the type parameter for component state from `any` to `{}`? This means I can no longer use `void` when my component has no state (but is not a stateless functional component).

**Comment by vsiao** [comment]
@seansfkelley for type safety, mostly (I don't want users to accidentally abuse props/state typed as `any`), but your wanting to pass `void` seems reasonable to me. Can you make a pull request for the changes that you need?

**Comment by JKillian** [comment]
On the `children` issue, the lifecycle methods seem incorrect now:

```
componentWillReceiveProps?(nextProps: P, nextContext: any): void;
```

 vs.

```
props: P & { children?: ReactNode };
```

**Comment by vsiao** [comment]
That's true. I guess we should `extend ComponentLifecycle<P & { children?: ReactNode }, S>`?

**Comment by JKillian** [comment]
@vsiao That's what I was thinking, unless there's some snazzier way to do it with a type alias

**Pull request merged**